vscode/src/gh-copilot/qsharpTools.ts

```diff
@@ -0,0 +1,289 @@
+   */
+  async runResourceEstimator(input: {
+    filePath: string;
+    qubitTypes?: string[];
+    errorBudget?: number;
+  }): Promise<
+    ProjectInfo & {
+      estimates?: object[];
+      message: string;
+    }
+  > {
+    const program = await this.getProgram(input.filePath);
+    const programConfig = program.programConfig;
+
+    const project = {
+      name: programConfig.projectName,
+      targetProfile: programConfig.profile,
+    };
+
+    try {
+      const qubitTypes = input.qubitTypes ?? ["qubit_gate_ns_e3"];
+      const errorBudget = input.errorBudget ?? 0.001;
+
+      const estimates = await resourceEstimateTool(
+        this.extensionUri,
+        programConfig,
+        qubitTypes,
+        errorBudget,
+      );
+
+      return {
+        project,
+        estimates,
+        message: "Results are displayed in the resource estimator panel.",
+      };
+    } catch (e) {
+      throw new CopilotToolError(
+        "Failed to run resource estimator: " +
+          (e instanceof Error ? e.message : String(e)),
+      );
+    }
+  }
+
+  private async getProgram(filePath: string) {
+    const docUri = vscode.Uri.file(filePath);
+
+    const doc = await vscode.workspace.openTextDocument(docUri);
+
+    const program = await getProgramForDocument(doc);
+    if (!program.success) {
+      throw new CopilotToolError(
+        `Cannot get program for the file ${filePath} . error: ${program.errorMsg}`,
+      );
+    }
+    return program;
+  }
+
+  private async runQsharp(
+    program: FullProgramConfig,
+    shots: number,
+    out: (message: string) => void,
+    resultUpdate: (
+      histogram: HistogramData,
+      failures: vscode.Diagnostic[],
+    ) => void,
+  ) {
+    let histogram: HistogramData | undefined;
+    const evtTarget = createDebugConsoleEventTarget((msg) => {
+      out(msg);
+    }, true /* captureEvents */);
+
+    // create a promise that we'll resolve when the run is done
+    let resolvePromise: () => void = () => {};
+    const allShotsDone = new Promise<void>((resolve) => {
+      resolvePromise = resolve;
+    });
+
+    evtTarget.addEventListener("uiResultsRefresh", () => {
+      const results = evtTarget.getResults();
+      const resultCount = evtTarget.resultCount(); // compiler errors come through here too
+      const buckets = new Map();
+      const failures = [];
+      for (let i = 0; i < resultCount; ++i) {
+        const key = results[i].result;
+        const strKey = typeof key !== "string" ? "ERROR" : key;
+        const newValue = (buckets.get(strKey) || 0) + 1;
+        buckets.set(strKey, newValue);
+        if (!results[i].success) {
+          failures.push(toVsCodeDiagnostic(results[i].result as VSDiagnostic));
+        }
+      }
+      histogram = {
+        buckets: Array.from(buckets.entries()),
+        shotCount: resultCount,
+      };
+      resultUpdate(histogram, failures);
+      if (shots === resultCount || failures.length > 0) {
+        // TODO: ugh
```

Comment:
??

Reply:
Oh no, I forgot about this :(  Behaviorally this should be ok, but it's a hack.

The problem was in the way the `run` function is designed. We can receive this result event *after* the function completes, but we do need to wait for all the results. We essentially never know when we're done. Originally I had this check `shots === resultCount` in here to understand we're done (i.e. we've run all the shots we set out to run), but that check would never be `true` if we had a compile error. So I had to add `failures.length > 0` check to cover the error case.

I need to revisit this.